### DISCUSSION TITLE
Question on Filter implementation

### DISCUSSION THREAD

**Discussion by hadamard-zweistein** [python-control/python-control#718] [Q&A]
I am interested in using the discrete transfer functions from the control library to filter signals and compute Fourier transforms.

For instance, given a discrete first order butterworth filter

```
from control import *

z = TransferFunction.z

c1 = 0.0196078431372549
c2 = -0.9607843137254902

Htest = c1*(1+z**(-1))/(1+c2*z**(-1))
```

How do I apply "Htest" on a signal "x" in order to obtain the filtered signal? Is there already a feature for that?

Likewise it would be very practical to use the transfer function of the Goertzel alg. to compute a fourier transform of a signal.

Thank you!

Hadamard-Zweistein

**Comment by bnavigator** [comment]
Hi,

You can use the [forced_response](https://python-control.readthedocs.io/en/0.9.1/generated/control.forced_response.html#control.forced_response) time domain simulation to filter a signal:

```python
R = forced_response(Htest, U=x)
y = R.y
```

But if you are dealing with SISO signals only, there may be no need to include python-control into your dependency stack. You can do everything with Scipy: https://scipy.github.io/devdocs/reference/generated/scipy.signal.lfilter.html#scipy.signal.lfilter:

```python
from scipy.signal import lfilter
b = c1*np.array([1, 1])
a = np.array([1, c2])
y = lfilter(b, a, x)
```

**Reply by hadamard-zweistein** [reply]
@bnavigator , thank you very much, both methods worked well!

For me, one advantage of the control version is the nice latex rendering of "H(z)" in the jupyter notebook

**Comment by hadamard-zweistein** [comment]
@bnavigator, one more question on complex coefficients.

Is there a way of using the forced_response method to compute Fourier transforms, for instance, to implement a Goertzel algorithm?

```
xn = np.array([1, 4, 9, 16, 25])

k = 1
N = len(xn) 

z = TransferFunction.z

H_goertzel = (1-np.exp(-2j*pi*k/N)*z**-1)/(1 -2*np.cos(2*pi*k/N)*z**-1+z**2) 
#H_goertzel
R = forced_response(H_goertzel, U=xn)

R.y
```

I get the warning, that the imaginary parts are ignored...

Thank you!

Hadamard-Zweistein

**Reply by bnavigator** [reply]
Your forgot a minus sign for the `z**(-2)`

```python
In [13]: H_goertzel = (1-np.exp(-2j*pi*k/N)*z**-1)/(1 -2*np.cos(2*pi*k/N)*z**-1+z**-2)

In [14]: H_goertzel
Out[14]: 
```
![image](https://user-images.githubusercontent.com/4623504/161750018-8ecea027-a3fe-4b0f-a076-e250e917fb2a.png)

```python
In [15]: H_goertzel.num
Out[15]: 
[[array([ 1.        +0.j        , -0.30901699+0.95105652j,
          0.        +0.j        ,  0.        +0.j        ,
          0.        +0.j        ])]]
In [16]: H_goertzel.den
Out[16]: [[array([ 1.        , -0.61803399,  1.        ,  0.        ,  0.        ])]]
```
The coefficients of the transfer function are complex. The resulting fourier transform needs to be complex. We do not support complex valued systems (yet). It has been on the TODO list forever. (see e.g. #484 and the issues linked there).

The LaTeX output is wrong here --> You found a bug: #719.

The scipy lfilter does what you want:

```python
In [17]: lfilter(H_goertzel.num[0][0], H_goertzel.den[0][0], xn)
Out[17]: 
array([ 1.         +0.j        ,  4.30901699 +0.95105652j,
        9.42705098 +4.39201132j, 14.73606798+10.3228644j ,
       19.73606798+17.20477401j])
```

**Reply by hadamard-zweistein** [reply]
@bnavigator, ok! Thank you! Looking forward to the complex analysis in control, there is a lot of stuff in classical textbooks, that could be implemented there. "lfilter" seems not so stable, the "control" version may be better.
For butterworth 1. order (real coeff.) control and scipy return good results.
For more sofistcated methods, like sdft (sliding discrete fourier transform), lfilter from scipy beheaves more unstable than a corresponding time domain implementation.
